File: ta_lib/strategies/confirm/src/eom.rs
use base::prelude::*;
use core::prelude::*;
use timeseries::prelude::*;
use volume::eom;

pub struct EomConfirm {
    source: SourceType,
    smooth: Smooth,
    period: usize,
}

impl EomConfirm {
    pub fn new(source: SourceType, smooth: Smooth, period: f32) -> Self {
        Self {
            source,
            smooth,
            period: period as usize,
        }
    }
}

impl Confirm for EomConfirm {
    fn lookback(&self) -> usize {
        self.period
    }

    fn filter(&self, data: &OHLCVSeries) -> (Series<bool>, Series<bool>) {
        let eom = eom(
            &data.source(self.source),
            data.high(),
            data.low(),
            data.volume(),
            self.smooth,
            self.period,
        );

        (eom.sgt(&ZERO_LINE), eom.slt(&ZERO_LINE))
    }
}

File: factor/confirm/eom.py
from dataclasses import dataclass

from core.models.parameter import Parameter, StaticParameter
from core.models.smooth import Smooth
from core.models.source import SourceType

from .base import Confirm, ConfirmType


@dataclass(frozen=True)
class EomConfirm(Confirm):
    type: ConfirmType = ConfirmType.Eom
    source_type: Parameter = StaticParameter(SourceType.HL2)
    smooth_type: Parameter = StaticParameter(Smooth.SMA)
    period: Parameter = StaticParameter(16.0)
